core/src/main/java/nl/nn/adapterframework/pipes/PGPPipe.java

```diff
@@ -0,0 +1,181 @@
+
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.IPipeLineSession;
+import nl.nn.adapterframework.core.PipeRunException;
+import nl.nn.adapterframework.core.PipeRunResult;
+import nl.nn.adapterframework.pgp.*;
+import nl.nn.adapterframework.stream.Message;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+import java.io.OutputStream;
+import java.security.Security;
+
+/**
+ * <p>Performs various PGP (Pretty Good Privacy) actions such as Encrypt, Sign, Decrypt, Verify.</p>
+ * <p>To use this pipe action parameter has to be set to one of the actions above.</p>
+ * <p>
+ * <br/><strong>Encrypt:</strong>
+ * <p>
+ * Requires the publicKey to be set to recipients public key,
+ * and recipients to be set to recipients email addresses.
+ * </p>
+ * <p>
+ * <br/><strong>Sign:</strong>
+ * <p>
+ * On top of the requirements for <i>Encrypt</i> action,
+ * signing requires senders to bet set for user's email;
+ * and secretKey & secretPassword to be set to private key's path and it's password
+ * (password is optional, if private key does not have protection).
+ * </p>
+ * <p>
+ * <br/><strong>Decrypt:</strong>
+ * <p>
+ * Requires secretKey and secretPassword to bet set to private key's path and it's password.
+ * Just like signing, password is not required, if private key does not have protection.
+ * </p>
+ * <p>
+ * <br/><strong>Verify:</strong>
+ * <p>
+ * On top of the requirements for <i>Decrypt</i> action,
+ * verification expects list of senders' email's and corresponding public keys.
+ * However, sender emails does not have to be set, and in that case,
+ * this pipe will only validate that someone signed the input.
+ * </p>
+ * <p>
+ * <br/>
+ * <p>
+ * <strong>Note:</strong> When secret key is required in any of the actions,
+ * the related public key should also be included in public keys.
+ * </p>
+ *
+ * <p>
+ * <strong>Note:</strong> For fields that require multiple input
+ * (such as recipients, senders and publicKey -in certain actions-)
+ * you can seperate multiple values with ";" (semicolon).
+ * </p>
+ */
+public class PGPPipe extends FixedForwardPipe {
+	/**
+	 * Action to be taken by pipe.
+	 * Available Actions:
+	 * <ul>
+	 *     <li>Encrypt: Encrypts the given input</li>
+	 *     <li>Sign: Encrypts and then signs the given input</li>
+	 *     <li>Decrypt: Decrypts the given input</li>
+	 *     <li>Verify: Decrypts and verifies the given input</li>
+	 * </ul>
+	 */
+	private String action;
+	/**
+	 * Emails of the recipients
+	 */
+	private String[] recipients;
+	/**
+	 * Emails of the senders. This will be used to verify that all the senders have signed the given message.
+	 * If not set, and the action is verify; this pipe will validate that at least one person has signed.
+	 *
+	 * For signing action, it needs to be set to the email that was used to generate the private key
+	 * that is being used for this process.
+	 */
+	private String[] senders;
```

Comment:
I'm afraid the name senders might confuse people... not sure if we should change this to something like verificationAddresses?

Reply:
Changed to verificationAddresses.